spec.bs

```diff
@@ -6,57 +6,882 @@ Status: CG-DRAFT
+    1. If |isConstructor| is false, throw a {{TypeError}}.
+    1. Let |prototype| be the result of running |operationCtor|'s [=object/[[GetPrototypeOf]]()=] method.
+    1. If |prototype| is not an [=object=], throw a {{TypeError}}.
+    1. Let |run| be the result of running [=GetMethod()=] with |prototype| and "`run`".
+    1. If |run| is undefined, throw a {{TypeError}}.
+    1. [=map/Set=] the value of |operationMap|[|name|] to |run|.
+  </div>
+
+The [=Shared Storage Database=] {#database}
+===========================================
+A [=user agent=] holds a <dfn>shared storage database</dfn>, which consists of a [=shared storage shed=] and provides methods to [=shared storage database/store an entry in the database|store=], [=shared storage database/retrieve an entry from the database|retrieve=], [=shared storage database/delete an entry from the database|delete=], [=shared storage database/clear all entries in the database|clear=], and [=shared storage database/purge expired entries from the database|purge expired=] data, and additional methods as below. The data in the [=shared storage database|database=] take the form of [=shared storage database/entry|entries=].
+
+Each <dfn for="shared storage database">entry</dfn> consists of a [=entry/key=] and a [=entry/value struct=].
+
+An [=shared storage database/entry=]'s <dfn for=entry>key</dfn> is a [=string=].
+
+[=User agents=] may specify the <dfn for=key>maximum length</dfn> of a [=entry/key=].
+
+Since [=entry/keys=] are used to organize and efficiently retrieve [=shared storage database/entry|entries=], [=entry/keys=] must appear at most once on any given [=shared storage shelf=] in a [=shared storage database|database=]'s [=shared storage shed=].
+
+An [=shared storage database/entry=]'s <dfn for=entry>value struct</dfn> is a [=struct=] composed of [=string=] <dfn for="value struct">value</dfn> and timestamp <dfn for="value struct">last updated</dfn>.
+
+[=User agents=] may specify the <dfn for=value>maximum length</dfn> of a [=value struct/value=].
+
+[=User agents=] may specify a <dfn>default entry lifetime</dfn>, the default duration between when an [=shared storage database/entry=] is [=shared storage database/store an entry in the database|stored=] and when it expires.
+
+A <dfn>shared storage database queue</dfn> for a [=/browsing context=] is a [=parallel queue=] on which the each of the [=shared storage database=]'s methods run when calls to them are initiated from that [=/browsing context=].
+
+  ## The [=Shared Storage Database|Database=] Algorithms ## {#database-algorithms}
+
+  <div algorithm>
+    To <dfn for="shared storage database">store an entry in the database</dfn>, given a [=/browsing context=] |context|, a [=entry/key=] |key|, and a [=value struct/value=] |value|, [=enqueue the following steps=] on |context|'s associated [=shared storage database queue=]:
+
+    1. If |key| is not a [=string=] or exceeds the [=key/maximum length=], return failure.
```

Comment:
Is "return failure" a standard term? I think this algorithm should return either true or false (ditto for others), and in the caller (e.g. worklet), check the result, and resolve or reject the promise accordingly.

Reply:
Yes, it's what the Storage Standard uses. See https://storage.spec.whatwg.org/#model and e.g. https://storage.spec.whatwg.org/#obtain-a-storage-shelf.

I will continue to use "failure" for the storage monkey patch, but changed to true/false or null/undefined elsewhere as applicable.